### PR TITLE
fix: Remove server closing of websocket connections.

### PR DESCRIPTION
**Pull request by SandPod**
Closes: #2952

The Serverpod streaming protocol supported closing the connection from the Server side if no open method stram connections were detected. However, this made it impossible for the client to synchronize its state since a connection might have been established and closed from the Server while the client was waiting for input.

This PR removes the server's possibility to close the WebSocket connections and makes the client authoritative around the lifecycle of WebSockets.

Note that the server can still close method streaming connections, for example, when a streaming endpoint has returned. It is simply not allowed to close the WebSocket connection as a result of this.

#### Test fixes

Some tests were removed as they validated the server's possibility of closing WebSocket connections. Other tests used the closing of the WebSocket connection to indicate that the streaming endpoint had returned.

These tests are now modified to rely on the session object or a StreamController, depending on what suited the test case best,  in the endpoint to provide that validation.

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - Client always closed WebSocket connections as well. Allowing the server to also close the connection was simply redundant.

### CODE REVIEW

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      'when calling an endpoint multiple times in quick succession '
+      'then all streams complete successfully', () async {
+    var client = Client(serverUrl);
+    var numMessages = 1000;
+    var stream;
+    for (var i = 0; i < 10; i++) {
+      stream = client.methodStreaming.intStreamFromValue(numMessages);
+    }
+
+    await expectLater(stream.last, completes);
```
Comment (christerswahn) [inline tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:19]: Suggestion: Verify all the streams complete normally with emitsDone matcher

File: tests/serverpod_test_server/lib/src/endpoints/method_streaming.dart [hunk 2]
```diff
@@ -14,15 +14,19 @@ class MethodStreaming extends Endpoint {
...
     Session session,
     int millisecondsDelay,
-  ) async* {
-    int i = 0;
-    while (true) {
-      await Future.delayed(Duration(milliseconds: millisecondsDelay));
-      yield i++;
-    }
+  ) {
+    var controller = StreamController<int>();
```
Comment (christerswahn) [inline tests/serverpod_test_server/lib/src/endpoints/method_streaming.dart:22]: Question: Is the controller "middle man" needed here to "keep it alive", as opposed to just returning result of Stream.periodic()?

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  });
+}
```
Comment (christerswahn) [inline tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:21]: Suggestion: Also add a test case where a bunch of streams from the same endpoint method are opened, and then read, paused, and resumed, in order to check pause-resume behavior with multiple streams open. (Unless that exists elsewhere already?)

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart [hunk 4]
```diff
@@ -0,0 +1,21 @@
+@Timeout(Duration(minutes: 5))
```
Comment (hampuslavin) [inline tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:1]: Question: How come this needs a 5 minute timeout?

**Review by christerswahn — COMMENTED** [review]
Looking forward to getting this in Ground Control! 😄

File: tests/serverpod_test_server/lib/src/endpoints/method_streaming.dart (hunk 2 above)
Comment (hampuslavin) [reply tests/serverpod_test_server/lib/src/endpoints/method_streaming.dart:22]: This is to enable the test to hook into the stream's lifecycle and assert that it was indeed cancelled in the end.

**Review by hampuslavin — COMMENTED** [review]
Nothing to add, just a small question 👍

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart (hunk 1 above)
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:19]: Good suggestion! Will fix! 👍

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart (hunk 4 above)
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:1]: This is not required, will remove it!

File: tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart (hunk 3 above)
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/open_stream_test.dart:57]: Good suggestion, added! 👍

**Review by hampuslavin — APPROVED** [review]
🏎️

**Pull request merged**